### ISSUE TITLE
Timeout does not work with session created with server

### ISSUE THREAD

**Issue by agnusmaximus** [tensorflow/tensorflow#6472]
### Environment info
Operating System: Ubuntu

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`): CPU only

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)  

c32663d119a4cbd9d4926d4eed2a0705f03a57f9

2. The output of `bazel version` 

Extracting Bazel installation...
Build label: 0.4.3
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Dec 22 12:31:25 2016 (1482409885)
Build timestamp: 1482409885
Build timestamp as int: 1482409885

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from datetime import datetime
import os.path
import time

import numpy as np
import random
import tensorflow as tf

import signal
import sys
import os

from tensorflow.python.framework import ops
from tensorflow.python.ops import variables
from tensorflow.python.ops import state_ops
from tensorflow.python.ops import logging_ops
from tensorflow.python.client import timeline

from tensorflow.python.ops import data_flow_ops

# (python test 1 worker > /tmp/output 2>&1 &) && (python test 0 ps > /tmp/output2 2>&1 &) && python test 0 worker

try:

    index, job = int(sys.argv[1]), sys.argv[2]


    ps_hosts = 'localhost:1234'
    worker_hosts = 'localhost:1235,localhost:1236'
    ps_hosts = ps_hosts.split(",")
    worker_hosts = worker_hosts.split(",")

    cluster_spec = tf.train.ClusterSpec({'ps' : ps_hosts, 'worker' : worker_hosts})
    server = tf.train.Server(
	{'ps': ps_hosts,
         'worker': worker_hosts},
	job_name=job,
	task_index=index)

    a = data_flow_ops.FIFOQueue(-1, tf.float32)
    b = a.dequeue()
    sv = tf.train.Supervisor(is_chief=True, logdir='./test_logdir')
    cfg = tf.ConfigProto(
	allow_soft_placement=True,
	log_device_placement=False)
    sess = sv.prepare_or_wait_for_session(server.target, config=cfg)
    print("Testing timeout")
    opt = tf.RunOptions(timeout_in_ms=50)
    sess.run(b, options=opt)
except tf.errors.DeadlineExceededError:
    print("Successfully timed out!")
```

Above I create a cluster and get a session from the server. I then try to dequeue an empty FIFOQueue with a timeout of 1 second. For some reason, on the latest commits of the tensorflow source, this does not successfully time out. This bug does not happen for the released versions of tensorflow (.11, .12).

To run:
`(python script.py 1 worker > /tmp/output 2>&1 &) && (python script.py 0 ps > /tmp/output2 2>&1 &) && python script.py 0 worker`

where script.py is the above example.

**Comment by mrry** [comment]
This looks like a bug in the (recently added) in-process shim for contacting the master without performing a full RPC. I'm testing a fix right now.

**Comment by mrry** [comment]
This should be fixed at HEAD now that #6530 is merged. See e1362157e33a6fd6a66caefd083bd4afc27dc030 for the actual fix.
